### DISCUSSION TITLE
TaskGraph or TaskSchedule?

### DISCUSSION THREAD

**Discussion by zealbell** [beehive-lab/TornadoVM#223] [Q&A] [answered]
In my maven project (since I am trying to replicate some examples from the **TornadoVM** repo), I have the following dependencies added `tornado-api` and `tornado-matrices` versions 0.14.1 but I’m getting IDE errors about `import uk.ac.manchester.tornado.api.TaskGraph`

I took a look at the *tornado-api-0.14.1.jar* from my project, lo and behold TaskGraph was nowhere to be found in it. I compared contents of the jar with **TornadoVM/tornado-api** and then I spotted a lookalike instead: “TaskSchedule”. 

I tried to make things work by simply swapping names in my project… currently everything is falling apart. Please Help🙏🏾

**Comment by jjfumero** [comment]
The `TaskGraph` is new only if you get the latest version from the `master` or `develop` branch. It is still work-in-progress. If you are using the official release (e.g., `0.14.1`, you won't see the `TaskGraph` yet. You simply use the `TaskSchedule` and the calls associated with it.
We highly recommend checking the examples package of TornadoVM to get an idea about how the APIs are called. 

Having said this, we are working on a new release, hopefully by the beginning 2023, with `TaskGraph` and a set of new interfaces. Just to be aware of this. We will also release a document that explains the migration of existing code to the new APIs and new documentation about the new APIs.

**Reply by zealbell** [reply]
That makes a lot of sense, although yesterday with the [ArrayMultiplyAdd](https://github.com/beehive-lab/TornadoVM/blob/master/tornado-examples/src/main/java/uk/ac/manchester/tornado/examples/arrays/ArrayMultiplyAdd.java) example  I wasn't able to get it to work by using `TaskSchedule`, some API calls did not tally. Is there any chance I get my hands on older examples with `TaskSchedule` (Say an older version of the ArrayMultiplyAdd.java class)

**Reply by jjfumero** [reply]
No, if you want to have those fixes. We strongly suggest using the latest version from the `master` branch.

**Reply by zealbell** [reply]
in my maven project? how do I go about doing that? (I have the latest version from the master branch)

**Reply by jjfumero** [reply]
You need to download the TornadoVM source code and compile it. Then, in your maven project you use:

```xml
<dependencies>   
      <dependency>
         <groupId>tornado</groupId>
         <artifactId>tornado-api</artifactId>
         <version>0.15-dev</version>
      </dependency>

      <dependency>
         <groupId>tornado</groupId>
         <artifactId>tornado-matrices</artifactId>
         <version>0.15-dev</version>
      </dependency>
</dependencies>   
```

With this, you get access to the latest version of the TornadoVM APIs and bug fixes in your code.

**Reply by zealbell** [reply]
thanks

**Reply by jjfumero** [reply]
No problem. Feel free to ask if you have questions.

**Reply by zealbell** [reply]
Hi, everything builds well but I'm getting a runtime exception thrown (whenever I run the example):

```java
Exception in thread "main" java.lang.NullPointerException
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:375)
	at uk.ac.manchester.tornado.api.runtime.TornadoAPIProvider.loadScheduleRuntime(TornadoAPIProvider.java:62)
	at uk.ac.manchester.tornado.api.TaskGraph.<init>(TaskGraph.java:85)
	at org.example.ArrayMultiplyAdd.main(ArrayMultiplyAdd.java:58)
```

**Reply by jjfumero** [reply]
Can you be more specific? Are you running TornadoVM through the `tornado` command? Do you have all the `env` variables needed?

https://github.com/beehive-lab/TornadoVM/blob/master/tornado-assembly/src/docs/1_INSTALL.md#installation

For example, in my case:

```bash
# source.sh file content 
export JAVA_HOME=/home/juan/manchester/tornado/tornado/etc/dependencies/TornadoVM-OpenJDK17/jdk-17.0.1
export PATH=/home/juan/manchester/tornado/tornado/bin/bin:$PATH
export TORNADO_SDK=/home/juan/manchester/tornado/tornado/bin/sdk
export CMAKE_ROOT=/home/juan/manchester/tornado/tornado/etc/dependencies/TornadoVM-OpenJDK17/cmake-3.22.1-linux-x86_64/
export TORNADO_ROOT=/home/juan/manchester/tornado/tornado 
```

**Reply by zealbell** [reply]
No, not through the `tornado` command. I have a standalone tornado-api maven project where I have  a copy of the  [ArrayMultiplyAdd](https://github.com/beehive-lab/TornadoVM/blob/master/tornado-examples/src/main/java/uk/ac/manchester/tornado/examples/arrays/ArrayMultiplyAdd.java). I built asper your help earlier asper:

```bash
cd /home/linker/TornadoVM
$ . source.sh
$ make graal-jdk-11-plus BACKEND=ptx
```
my source:

```bash
export JAVA_HOME=/home/linker/TornadoVM/etc/dependencies/TornadoVM-GraalJDK17/graalvm-ce-java17-22.2.0
export PATH=/home/linker/TornadoVM/bin/bin:$PATH
export TORNADO_SDK=/home/linker/TornadoVM/bin/sdk
export CMAKE_ROOT=/home/linker/TornadoVM/etc/dependencies/TornadoVM-GraalJDK17/cmake-3.22.1-linux-x86_64/
export TORNADO_ROOT=/home/linker/TornadoVM 
```

**Reply by jjfumero** [reply]
The error you are getting is related to missing flags when running `java`. We suggest running Tornado applications using the `tornado` command. The `tornado` command is just an alias to `java` and all flags that enable TornadoVM execution. 

If you do not want to use the `tornado` command, you can inspect all flags needed by running:

```bash
$ tornado --printJavaFlags
```

Example of output:
```bash
$ tornado --printJavaFlags
/home/juan/tornado/TornadoVM/etc/dependencies/TornadoVM-OpenJDK17/jdk-17.0.1/bin/java -server -XX:-UseCompressedOops -XX:+UnlockExperimentalVMOptions -XX:+EnableJVMCI -Djava.library.path=/home/juan/tornado/TornadoVM/bin/sdk/lib  --module-path .:/home/juan/tornado/TornadoVM/bin/sdk/share/java/tornado -Dtornado.load.api.implementation=uk.ac.manchester.tornado.runtime.tasks.TornadoTaskGraph -Dtornado.load.runtime.implementation=uk.ac.manchester.tornado.runtime.TornadoCoreRuntime -Dtornado.load.tornado.implementation=uk.ac.manchester.tornado.runtime.common.Tornado -Dtornado.load.device.implementation.opencl=uk.ac.manchester.tornado.drivers.opencl.runtime.OCLDeviceFactory -Dtornado.load.device.implementation.ptx=uk.ac.manchester.tornado.drivers.ptx.runtime.PTXDeviceFactory -Dtornado.load.device.implementation.spirv=uk.ac.manchester.tornado.drivers.spirv.runtime.SPIRVDeviceFactory -Dtornado.load.annotation.implementation=uk.ac.manchester.tornado.annotation.ASMClassVisitor -Dtornado.load.annotation.parallel=uk.ac.manchester.tornado.api.annotations.Parallel  --upgrade-module-path /home/juan/tornado/TornadoVM/bin/sdk/share/java/graalJars -XX:+UseParallelGC  @/home/juan/tornado/TornadoVM/bin/sdk/etc/exportLists/common-exports @/home/juan/tornado/TornadoVM/bin/sdk/etc/exportLists/opencl-exports @/home/juan/tornado/TornadoVM/bin/sdk/etc/exportLists/spirv-exports --add-modules ALL-SYSTEM,tornado.runtime,tornado.annotation,tornado.drivers.common,tornado.drivers.opencl
```

These flags hold for both using the latest TornadoVM Maven files and using the latest build from the `master` or `develop` branch.

**Reply by zealbell** [reply]
Hi, Just got back on my feet (I fell sick). I only wish to run (or use) tornado (api) from my standalone maven project (not via the cloned `master` branch), currently this is what I get:

```
/home/linker/TornadoVM/etc/dependencies/TornadoVM-GraalJDK17/graalvm-ce-java17-22.2.0/bin/java -server -XX:-UseCompressedOops -XX:+UnlockExperimentalVMOptions -XX:+EnableJVMCI -Djava.library.path=/home/linker/TornadoVM/bin/sdk/lib  --module-path .:/home/linker/TornadoVM/bin/sdk/share/java/tornado -Dtornado.load.api.implementation=uk.ac.manchester.tornado.runtime.tasks.TornadoTaskGraph -Dtornado.load.runtime.implementation=uk.ac.manchester.tornado.runtime.TornadoCoreRuntime -Dtornado.load.tornado.implementation=uk.ac.manchester.tornado.runtime.common.Tornado -Dtornado.load.device.implementation.opencl=uk.ac.manchester.tornado.drivers.opencl.runtime.OCLDeviceFactory -Dtornado.load.device.implementation.ptx=uk.ac.manchester.tornado.drivers.ptx.runtime.PTXDeviceFactory -Dtornado.load.device.implementation.spirv=uk.ac.manchester.tornado.drivers.spirv.runtime.SPIRVDeviceFactory -Dtornado.load.annotation.implementation=uk.ac.manchester.tornado.annotation.ASMClassVisitor -Dtornado.load.annotation.parallel=uk.ac.manchester.tornado.api.annotations.Parallel  -XX:+UseParallelGC  @/home/linker/TornadoVM/bin/sdk/etc/exportLists/common-exports @/home/linker/TornadoVM/bin/sdk/etc/exportLists/ptx-exports --add-modules ALL-SYSTEM,tornado.runtime,tornado.annotation,tornado.drivers.common,tornado.drivers.ptx 
```

What flags are not correctly set?

**Reply by jjfumero** [reply]
They are correct if you have installed the drivers for CUDA, OpenCL and SPIRV. Do you have support for these backends in your current setup?

**Reply by zealbell** [reply]
Yes, I do (with the exception of SPIRV). I tried changing my backend to `opencl`. I still get the same error

**Reply by jjfumero** [reply]
What's the error you are getting? Which version of TornadoVM are you using, and which OS are you using? Note also that Java flags are different per backend.

**Reply by zealbell** [reply]
> Hi, everything builds well but I'm getting a runtime exception thrown (whenever I run the example):
> 
> ```java
> Exception in thread "main" java.lang.NullPointerException
> 	at java.base/java.lang.Class.forName0(Native Method)
> 	at java.base/java.lang.Class.forName(Class.java:375)
> 	at uk.ac.manchester.tornado.api.runtime.TornadoAPIProvider.loadScheduleRuntime(TornadoAPIProvider.java:62)
> 	at uk.ac.manchester.tornado.api.TaskGraph.<init>(TaskGraph.java:85)
> 	at org.example.ArrayMultiplyAdd.main(ArrayMultiplyAdd.java:58)
> ```

Ubuntu 22.0.4, Tornado 0.15-dev, Backend=ptx

**Reply by jjfumero** [reply]
That error means you are not providing the location of the PTX API Provider .

Using my configuration:

```bash
-server -XX:-UseCompressedOops -XX:+UnlockExperimentalVMOptions -XX:+EnableJVMCI -Djava.library.path=/home/juan/manchester/tornado/tornado/bin/sdk/lib  --module-path .:/home/juan/manchester/tornado/tornado/bin/sdk/share/java/tornado -Dtornado.load.api.implementation=uk.ac.manchester.tornado.runtime.tasks.TornadoTaskGraph -Dtornado.load.runtime.implementation=uk.ac.manchester.tornado.runtime.TornadoCoreRuntime -Dtornado.load.tornado.implementation=uk.ac.manchester.tornado.runtime.common.Tornado -Dtornado.load.device.implementation.opencl=uk.ac.manchester.tornado.drivers.opencl.runtime.OCLDeviceFactory -Dtornado.load.device.implementation.ptx=uk.ac.manchester.tornado.drivers.ptx.runtime.PTXDeviceFactory -Dtornado.load.device.implementation.spirv=uk.ac.manchester.tornado.drivers.spirv.runtime.SPIRVDeviceFactory -Dtornado.load.annotation.implementation=uk.ac.manchester.tornado.annotation.ASMClassVisitor -Dtornado.load.annotation.parallel=uk.ac.manchester.tornado.api.annotations.Parallel  --upgrade-module-path /home/juan/manchester/tornado/tornado/bin/sdk/share/java/graalJars -XX:+UseParallelGC  @/home/juan/manchester/tornado/tornado/bin/sdk/etc/exportLists/common-exports @/home/juan/manchester/tornado/tornado/bin/sdk/etc/exportLists/ptx-exports --add-modules ALL-SYSTEM,tornado.runtime,tornado.annotation,tornado.drivers.common,tornado.drivers.ptx 
```

**Reply by zealbell** [reply]
Please how am I supposed to do that, I didn’t come about anything relating to this in the docs I went through.

**Reply by jjfumero** [reply]
You run the following command:

```bash
tornado --printJavaFlags
```

**Reply by jjfumero** [reply]
I assume you are not running from the terminal, but rather from the IDE. In this case, you need to pass those parameters when you run your program so you need to configure the IDE to use those flags per application.

**Reply by zealbell** [reply]
> Hi, Just got back on my feet (I fell sick). I only wish to run (or use) tornado (api) from my standalone maven project (not via the cloned `master` branch), currently this is what I get:
> 
> ```
> /home/linker/TornadoVM/etc/dependencies/TornadoVM-GraalJDK17/graalvm-ce-java17-22.2.0/bin/java -server -XX:-UseCompressedOops -XX:+UnlockExperimentalVMOptions -XX:+EnableJVMCI -Djava.library.path=/home/linker/TornadoVM/bin/sdk/lib  --module-path .:/home/linker/TornadoVM/bin/sdk/share/java/tornado -Dtornado.load.api.implementation=uk.ac.manchester.tornado.runtime.tasks.TornadoTaskGraph -Dtornado.load.runtime.implementation=uk.ac.manchester.tornado.runtime.TornadoCoreRuntime -Dtornado.load.tornado.implementation=uk.ac.manchester.tornado.runtime.common.Tornado -Dtornado.load.device.implementation.opencl=uk.ac.manchester.tornado.drivers.opencl.runtime.OCLDeviceFactory -Dtornado.load.device.implementation.ptx=uk.ac.manchester.tornado.drivers.ptx.runtime.PTXDeviceFactory -Dtornado.load.device.implementation.spirv=uk.ac.manchester.tornado.drivers.spirv.runtime.SPIRVDeviceFactory -Dtornado.load.annotation.implementation=uk.ac.manchester.tornado.annotation.ASMClassVisitor -Dtornado.load.annotation.parallel=uk.ac.manchester.tornado.api.annotations.Parallel  -XX:+UseParallelGC  @/home/linker/TornadoVM/bin/sdk/etc/exportLists/common-exports @/home/linker/TornadoVM/bin/sdk/etc/exportLists/ptx-exports --add-modules ALL-SYSTEM,tornado.runtime,tornado.annotation,tornado.drivers.common,tornado.drivers.ptx 
> ```
> 
> What flags are not correctly set?

I did this earlier (`$tornado --printJavaFlags`) and you said my flags were ok

**Reply by jjfumero** [reply]
As flags in the IDE you can't use the `/home/linker/TornadoVM/etc/dependencies/TornadoVM-GraalJDK17/graalvm-ce-java17-22.2.0/bin/java`. THis is not a Java flag, This is the path to your Java installation.

**Reply by zealbell** [reply]
I am not sure I quite follow with your last message. Are you hinting that after I run the `$ tornado --printJavaFlags` command, I have to manually set my IDE flags to be exactly as the output I get?

**Reply by jjfumero** [reply]
You need to configure your IDE to launch your applications using the Java flags displayed from the command: `tornado --printJavaFlags` without the path to Java (obviously). 

Can you share how are you running your Java programs in your editor? Which IDE are you using? 

Otherwise, you can use the command line to run your TornadoVM programs.

**Reply by zealbell** [reply]
I am using Intellij, I am running my project by building with maven. I thought the `tornado --printJavaFlags` output were flags currently set automatically by the IDE in sync with the maven build.

**Comment by zealbell** [comment, marked as answer]
Hi Juan, Big Thanks🔥. My card is spinning⚡️ currently, I went over everything you said again, everything now works. I am just back to post about my journey so far on this thread to help others because I feel most of the issues I encountered could have been avoided if there was some blog or guide about creating a standalone tornado maven project in Intellij. 

**INSTALLATION**
```bash
$ cd /home/linker
$ git clone https://github.com/beehive-lab/TornadoVM.git
$ cd TornadoVM
$ ./scripts/tornadoVMInstaller.sh --graal-jdk-17 --opencl --ptx
```

**SETUP**

- Intellij: creation of 'Tornadoes' as a [maven project](https://github.com/beehive-lab/TornadoVM#5-how-to-use-tornadovm-in-your-projects) (i.e. /home/linker/Tornadoes | current version in my `pom.xml` is 0.15-dev instead) with `MatrixMultiplication2D.java` example from the **TornadoVM** `master` branch.


**LAUNCH**

- Open the project TornadoVM


**BUILD (ONETIME/SETUP)** [ *previously: BUILD (EVERYTIME/LAUNCH)* ]
```
$ cd /home/linker/TornadoVM
$ . source.sh
$ make graal-jdk-11-plus BACKEND=ptx 
```


**CONFIGURATION (ONETIME/PROGRAM)**

```bash
$ tornado --printJavaFlags
``` 

- Intellij --> Run --> Edit Configurations --> `<Your Program>` --> Modify Options --> Add VM Options >> VM Options | paste `$ tornado --printJavaFlags`  output without the java path (i.e. starting from `-server...`) --> Apply --> Ok


**TESTING**
my ❤️ly output:
```bash
Computing MxM of 512x512
Single Threaded CPU Execution: 1.18 GFlops, Total time = 227 ms
TornadoVM Execution on GPU (Accelerated): 268.44 GFlops, Total Time = 1 ms #didn't think tornado would be this fast... nice!
Speedup: 227.0x
Verification true
```
after running `/home/linker/Tornadoes/src/main/java/org/example/MatrixMultiplication2D.java`

**Reply by jjfumero** [reply]
To clarify, you do not need to build TornadoVM every time, but there are a few exceptions:

1. You modify the TornadoVM project. 
2. You want to access another backend (e.g., SPIR-V).  In this case, you recompile TornadoVM with the desired backends. 

Otherwise, it just needs to build once, as with any other Java project.

**Reply by zealbell** [reply]
Oh! alright then. thankie😊
